### ISSUE TITLE
PIV PKCS15 emulator always fetches entire card

### ISSUE THREAD

**Issue by darconeous** [OpenSC/OpenSC#377]
The PKCS15 emulator for PIV-based cards always pre-fetches all objects on the card (including certificates), no matter what operation is being performed. This makes performing authentication operations on such cards take many seconds if the card has several certificates. This behavior makes the optional object caching mechanism in PKCS15 completely useless.

I implemented a quick fix for this behavior in 0ffad3f3bb9c893ab5a06998c7c7f0d571e7918d which improves the situation when `use_file_cache` is set, but it still seems like there is significant room for improvement in terms of performance.

I'm happy to offer 0ffad3f3bb9c893ab5a06998c7c7f0d571e7918d as a pull request, but I'm not sure if it is the right solution to this problem. It seems like, no matter what the file cache situation may be, we shouldn't be automatically pre-fetching all certificates on the card.

**Comment by dengert** [comment]
On 2/12/2015 6:53 PM, Robert Quattlebaum wrote:
> The PKCS15 emulator for PIV-based cards always pre-fetches all objects on the card (including certificates), no matter what operation is being performed. This makes performing authentication
> operations on such cards take many seconds if the card has several certificates. This behavior makes the optional object caching mechanism in PKCS15 completely useless.
>

The card does not have public or private key objects. There is no way to determine
the type of keys or their size other then to read the certificate, parse the
certificate to get the SPKI and determine the type of key, RSA or ECC and the size of the key.

See NIST 800-73-3_PART 1,
  APPENDIX C.1 "PIV Algorithm Identifier Discovery for Asymmetric Cryptograph Authentication"

 From a PKCS#11 perspective any query for certificate objects, public key objects
or private key objects will cause the certificates to be read.

> I implemented a quick fix for this behavior in 0ffad3f <https://github.com/OpenSC/OpenSC/commit/0ffad3f3bb9c893ab5a06998c7c7f0d571e7918d> which improves the situation when |use_file_cache| is set, but
> it still seems like there is significant room for improvement in terms of performance.

The card-piv.c will cache objects in memory from the card on its own. It does not use the pkcs15 cache,
as this is not a pkcs15 card.

You might also notice that the certificate is contained within a certificate object,
and it might be gziped as well. Part of the card-piv.c caching it to cache both the
certificate object, and the unzipped certificate.

>
> I'm happy to offer 0ffad3f <https://github.com/OpenSC/OpenSC/commit/0ffad3f3bb9c893ab5a06998c7c7f0d571e7918d> as a pull request, but I'm not sure if it is the right solution to this problem. It seems
> like, no matter what the file cache situation may be, we shouldn't be automatically pre-fetching all certificates on the card.
>

Believe me I complained to NIST about this years ago.


Aoso I don't understand you patch. It does not look like it would do anything.

I also don't see how you are using the pkcsc15 cache. Are you using pkcs15-tool.c to cache the card contents?
What will pkcs15-tool cache?

The PIV card is not a pkcs15 card, and emulates much of the pkcs15 functionality
and it may be bypassing parts of the code that use the file cache.

The card can also be used for login before any home directory is accessible. In that case
or a file not be found, so the card will still get read.

In addition the CHUID object is read from the card to derive the card serial number,
which is used in the cache names. So you can't cache the CHUID.

> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/377>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by darconeous** [comment]
The changes simply allow the use of the file cache if it is enabled. More specifically, it allows the file cache to actually be created properly in the first place. Once the cache files are created properly then they are used.

If `use_file_cache` is not enabled, then the behavior continues as normal, fetching all certificates every time. The big win is that if it can fetch a certificate from a file instead of from the card, it will do so.

Long-term, it would be great to also add caching of some of the other objects as well (except, obviously, the CHUID object), but just fixing the caching of the certificates alone seems to shave off several seconds.

In my application I'm putting together a physical access control system that uses a Raspberry PI, an NFC reader, and an electric door strike... So caching files isn't a problem because no one is "logging in", but waiting any longer than a second or so to open the door is a non-starter.

**Comment by dengert** [comment]
On 2/13/2015 12:06 PM, Robert Quattlebaum wrote:
> The changes simply allow the use of the file cache if it is enabled. More specifically, it allows the file cache to actually be created properly in the first place. Once the cache files are created
> properly then they are used.
>
> If |use_file_cache| is not enabled, then the behavior continues as normal, fetching all certificates every time. The big win is that if it can fetch a certificate from a file instead of from the card,
> it will do so.
>
> Long-term, it would be great to also add caching of some of the other objects as well (except, obviously, the CHUID object), but just fixing the caching of the certificates alone seems to shave off
> several seconds.
>
> In my application I'm putting together a physical access control system that uses a Raspberry PI, an NFC reader, and an electric door strike... So caching files isn't a problem because no one is
> "logging in", but waiting any longer than a second or so to open the door is a non-starter.

Is this a system for your own use, or are you going to market this?

Are you using PIV cards officially issued in accordance with FIPS 201 standards by a US gov agency?
or are these going to be PIV-like cards?

Have you read NIST 800-73-3_part1 Appendix B?

Are you going to use the Card Authentication cert and matching 9E key with a Appendix B.1.5?

Do your user keep their card in a shield card holder to protect against NFC easy dropping?

Are you going to check signatures on the CHUID and verify certificates including using CRL or OSCP?

I ask because:
If its for your own use, I am reluctant to make the current changes you have propose. Maybe you should
find a different card. There are a lot of people out there using the gov issued PIV cards with
OpenSC, that expect the code to meet the NIST standards.

If this is a PIV like card that does not meet all the NIST requirements, in particular
  NIST 800-73-3 part 2 After table 2:

    "Note: Cryptographic protocols using private/secret keys requiring “PIN” security
     condition shall not be used on the contactless interface."

you are opening up yourself or your product to security risks, especially
if your users don't use card shields.

If you are not using the 9E key there are even more risks in using cached objects.

NFC for security has a lot of issues, which are apparent in the fact that PIV card users
are required to keep the card in a shield.

Even credit cards have switched from NFC to using readers, the speed passes are on their way
out, many credit cards that used to supported Blink, don't now, and newer cards have a chip
especially in Europe, with the chip-and-pin, and the US is getting ready for this as well.
Next time in a store with a credit card/debit card reader, note that all newer readers
also have a slot at the bottom for the card.

  Google for: credit card  EMV.

>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/377#issuecomment-74297969>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by dengert** [comment]
I am working on some mods for contactless access to PIV, that should speed up processing
by only trying to load the objects that are allowed in contactless mode.
I am also trying to accommodate not standard PIV like devices that violate this standard.
The opensc.conf would have to be configured to violate the contactless mode.

A missing piece at the moment is how to have pcscd to pass in the fact that a reader
contactless.

The security is really in the card, but knowing of it is a contactless reader would
improve performance.



On 2/13/2015 12:06 PM, Robert Quattlebaum wrote:
> The changes simply allow the use of the file cache if it is enabled. More specifically, it allows the file cache to actually be created properly in the first place. Once the cache files are created
> properly then they are used.
>
> If |use_file_cache| is not enabled, then the behavior continues as normal, fetching all certificates every time. The big win is that if it can fetch a certificate from a file instead of from the card,
> it will do so.
>
> Long-term, it would be great to also add caching of some of the other objects as well (except, obviously, the CHUID object), but just fixing the caching of the certificates alone seems to shave off
> several seconds.
>
> In my application I'm putting together a physical access control system that uses a Raspberry PI, an NFC reader, and an electric door strike... So caching files isn't a problem because no one is
> "logging in", but waiting any longer than a second or so to open the door is a non-starter.
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/377#issuecomment-74297969>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by darconeous** [comment]
> Is this a system for your own use, or are you going to market this?

My own use.

> Are you using PIV cards officially issued in accordance with FIPS 201 standards by a US gov agency?
or are these going to be PIV-like cards?

I intend to use cards that are designed to be compatible with 800-73-3 systems but have not been officially issued or certified for FIPS 201 compliance. (*UPDATE:* the reason why you asked this question became clear after I wrote this whole long post, so make sure you read the paragraph at the end that has *AH-HAH* in it)

> Have you read NIST 800-73-3_part1 Appendix B?

Yes, indeed I have.

> Are you going to use the Card Authentication cert and matching 9E key with a Appendix B.1.5?

Yes. This is the only key I can legitimately use via NFC.

> Do your user keep their card in a shield card holder to protect against NFC easy dropping?

Unlikely, but I would encourage it. However, such a snooper would only be able to retrieve the PIV CHUID, PIV CAC, NFC UID, and NFC ATS/ATR. The information contained in the certificate would only identify the card, not the person. So while a snooper would be able to identify if they have snooped that card before or not, they wouldn't be able to tell much more.

> Are you going to check signatures on the CHUID?

Probably not, since I would just be using it to retrieve a cached certificate.

>  Are you going to certificates including using CRL or OSCP?

Yes, absolutely. I'm not going to issue cards that I can't revoke, that would be silly.

> I ask because: If its for your own use, I am reluctant to make the current changes you have propose.

Why is my intended use have any relevance to you accepting a patch or not? If it's a good change, I would expect you to take it, and if it's a bad change, I'd expect you to not take it. Whether I'm implementing a commercial system or not shouldn't matter, and if it does then that doesn't bode well for the integrity of the project.

> Maybe you should find a different card.

Why?

> There are a lot of people out there using the gov issued PIV cards with OpenSC, that expect the code to meet the NIST standards.

If you think that I have suggested that the code should be changed in such a way as to make it not meet NIST standards, then I believe you have dramatically misunderstood what I am trying to do.

> If this is a PIV like card that does not meet all the NIST requirements, in particular NIST 800-73-3 part 2 After table 2:
>
> > "Note: Cryptographic protocols using private/secret keys requiring “PIN” security condition shall not be used on the contactless interface."
>
> you are opening up yourself or your product to security risks, especially if your users don't use card shields.

Which is exactly why this is *NOT* what I'm doing. I *never* said this was what I was trying to do.

> NFC for security has a lot of issues, which are apparent in the fact that PIV card users are required to keep the card in a shield.

They are asked to keep the card in a shield because government employees are potentially high-value targets. If someone can correlate a CHUID with a specific individual, then they could implement person-specific bombs and whatnot. As long as the underlying PIV app on the card is written properly (Only allowing the CHUID and CAC to be read/used when accessed via NFC) then this is the only security risk.

> Even credit cards have switched from NFC to using readers, the speed passes are on their way out, many credit cards that used to supported Blink, don't now, and newer cards have a chip especially in Europe, with the chip-and-pin, and the US is getting ready for this as well.
>Next time in a store with a credit card/debit card reader, note that all newer readers also have a slot at the bottom for the card.

This is an entirely different problem. Cite specific examples of reasons why NFC is being phased out of payment cards and why those reasons would apply to PIV cards.

--------------

My proposed process is:

1. User presents card to reader.
2. Reader reads ATR, compares to whitelist. If not, reject access.
3. Reader selects PIV application. If this fails, reject access.
4. Reader reads CHUID.
5. Reader checks to see if we have a certificate cached for it. If so, use it. If not, read out the certificate.
6. Send a challenge to the card for key '9E' to verify that the card has the private key for the cert we have in hand.
    * If the challenge fails and the cert was read from a cache, obliterate the cache entry for that CHUID and goto 5. 
    * If the challenge fails and the cert was read directly from the card, reject access.
7. Determine if the certificate has access to open this door, checking revocations, expirations, chain-of-trust, etc.
    * If the verification fails and the cert was read from a cache, obliterate the cache entry for that CHUID and goto 5.
    * If the verification fails and the cert was read directly from the card, reject access.
8. Grant access and store the certificate for the given CHUID in the cache to improve the speed of the next access.

This process presumes that it is faster to fetch the CHUID than the CAC. If this isn't the case, or it turns out that the difference in timing isn't really perceptible, then I'd be fine throwing out the caching mechanism altogether.

However, when the PKCS15 emulator starts up, it *always* fetches *all* objects on the card, no matter what sort of operation I am performing.

Oh... Hey, wait a sec... *AH-HAH*! I think I see where the confusion is coming form. The [PIV app](https://developers.yubico.com/yubico-piv-tool/YubiKey_NEO_PIV_introduction.html) on the [YubiKey NEO](https://www.yubico.com/products/yubikey-hardware/yubikey-neo/) (which I'm using for prototyping and happens to have the other certs and keys present) does not differentiate between NFC and USB, allowing *ALL* operations to be performed via NFC (which is really unfortunate from a security standpoint). For a real, compliant PIV card (which is what I intend to use for this application), this wouldn't be a problem because the only objects that could be fetched would be the CHUID and the CAC. This would likely make the delay between credential presentation to the reader and the door unlocking much more acceptable in practice.

In any case, I don't think that allowing the built-in PKCS15 caching mechanism to work properly would break anything, for two reasons:

1. You must opt in to use the file cache by setting `use_file_cache`. By default it is turned off.
2. `use_file_cache` does not automatically cache objects. If you want a card to be cached, you must explicitly do so.

**Comment by darconeous** [comment]
> I am working on some mods for contactless access to PIV, that should speed up processing by only trying to load the objects that are allowed in contactless mode. I am also trying to accommodate not standard PIV like devices that violate this standard. The opensc.conf would have to be configured to violate the contactless mode.

This sounds great, especially if it could be configured to always treat PIV cards as "contactless" and only fetch the contactless subset. (In my application, this would be the case)

> A missing piece at the moment is how to have pcscd to pass in the fact that a reader contactless.

From a PCSC perspective, you'd probably have to introspect into the description of the reader and differentiate between contact and contactless readers (In the case where it supports both I guess you'd assume contact-only). Kinda yucky, but doable.

> The security is really in the card, but knowing of it is a contactless reader would improve performance.

Certainly better than nothing!

**Comment by dengert** [comment]
Good answers, Including the /AH-HAH/ at the end.

I reply to some of your other comments including performance please see:

https://github.com/dengert/OpenSC/compare/piv-contactless

There are two experimental commits in the piv-contactless branch.

The first tries to address the use of a contactless interface to a PIV card,
only trying to read the objects that are listed as readable in contactless mode.
It also tries to address the needs of non-approved cards like the YubiKey NEO
that treats the contact and contactless interfaces the same.
See the commit message on how to use the flags= in opensc.conf
Yubico people should also look at this.

Not obvious is how to test if a reader is contactless. I need to look at
pcsd and CCID features to see if OpenSC can detect if a reader is contact or
contactless. In the bigger picture, it is the responibility of the card to treat
contactless different then contact.

The second commit is to improve the reading of objects. A recent change in apdu.c
would cause a full read of an object, when the code used to work by reading
the first 8 bytes (or 256 bytes with  T=1) to get the length, then read the full object.
This is the cause of the double reads you reported.



On 2/20/2015 6:43 PM, Robert Quattlebaum wrote:
>     Is this a system for your own use, or are you going to market this?
>
> My own use.
>
>     Are you using PIV cards officially issued in accordance with FIPS 201 standards by a US gov agency?
>     or are these going to be PIV-like cards?
>
> I intend to use cards that are designed to be compatible with 800-73-3 systems but have not been officially issued or certified for FIPS 201 compliance. (/UPDATE:/ the reason why you asked this
> question became clear after I wrote this whole long post, so make sure you read the paragraph at the end that has /AH-HAH/ in it)
>
>     Have you read NIST 800-73-3_part1 Appendix B?
>
> Yes, indeed I have.
>
>     Are you going to use the Card Authentication cert and matching 9E key with a Appendix B.1.5?
>
> Yes. This is the only key I can legitimately use via NFC.
>
>     Do your user keep their card in a shield card holder to protect against NFC easy dropping?
>
> Unlikely, but I would encourage it. However, such a snooper would only be able to retrieve the PIV CHUID, PIV CAC, NFC UID, and NFC ATS/ATR. The information contained in the certificate would only
> identify the card, not the person. So while a snooper would be able to identify if they have snooped that card before or not, they wouldn't be able to tell much more.
>
>     Are you going to check signatures on the CHUID?
>
> Probably not, since I would just be using it to retrieve a cached certificate.
>
>     Are you going to certificates including using CRL or OSCP?
>
> Yes, absolutely. I'm not going to issue cards that I can't revoke, that would be silly.
>
>     I ask because: If its for your own use, I am reluctant to make the current changes you have propose.
>
> Why is my intended use have any relevance to you accepting a patch or not? If it's a good change, I would expect you to take it, and if it's a bad change, I'd expect you to not take it. Whether I'm
> implementing a commercial system or not shouldn't matter, and if it does then that doesn't bode well for the integrity of the project.
>
>     Maybe you should find a different card.
>
> Why?
>
>     There are a lot of people out there using the gov issued PIV cards with OpenSC, that expect the code to meet the NIST standards.
>
> If you think that I have suggested that the code should be changed in such a way as to make it not meet NIST standards, then I believe you have dramatically misunderstood what I am trying to do.
>
>     If this is a PIV like card that does not meet all the NIST requirements, in particular NIST 800-73-3 part 2 After table 2:
>
>         "Note: Cryptographic protocols using private/secret keys requiring “PIN” security condition shall not be used on the contactless interface."
>
>     you are opening up yourself or your product to security risks, especially if your users don't use card shields.
>
> Which is exactly why this is /NOT/ what I'm doing. I /never/ said this was what I was trying to do.
>
>     NFC for security has a lot of issues, which are apparent in the fact that PIV card users are required to keep the card in a shield.
>
> They are asked to keep the card in a shield because government employees are potentially high-value targets. If someone can correlate a CHUID with a specific individual, then they could implement
> person-specific bombs and whatnot. As long as the underlying PIV app on the card is written properly (Only allowing the CHUID and CAC to be read/used when accessed via NFC) then this is the only
> security risk.
>
>     Even credit cards have switched from NFC to using readers, the speed passes are on their way out, many credit cards that used to supported Blink, don't now, and newer cards have a chip especially
>     in Europe, with the chip-and-pin, and the US is getting ready for this as well.
>     Next time in a store with a credit card/debit card reader, note that all newer readers also have a slot at the bottom for the card.
>
> This is an entirely different problem. Cite specific examples of reasons why NFC is being phased out of payment cards and why those reasons would apply to PIV cards.
>
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
>
> My proposed process is:
>
>  1. User presents card to reader.
>  2. Reader reads ATR, compares to whitelist. If not, reject access.
>  3. Reader selects PIV application. If this fails, reject access.
>  4. Reader reads CHUID.
>  5. Reader checks to see if we have a certificate cached for it. If so, use it. If not, read out the certificate.
>  6. Send a challenge to the card for key '9E' to verify that the card has the private key for the cert we have in hand.
>       * If the challenge fails and the cert was read from a cache, obliterate the cache entry for that CHUID and goto 5.
>       * If the challenge fails and the cert was read directly from the card, reject access.
>  7. Determine if the certificate has access to open this door, checking revocations, expirations, chain-of-trust, etc.
>       * If the verification fails and the cert was read from a cache, obliterate the cache entry for that CHUID and goto 5.
>       * If the verification fails and the cert was read directly from the card, reject access.
>  8. Grant access and store the certificate for the given CHUID in the cache to improve the speed of the next access.
>
> This process presumes that it is faster to fetch the CHUID than the CAC. If this isn't the case, or it turns out that the difference in timing isn't really perceptible, then I'd be fine throwing out
> the caching mechanism altogether.
>
> However, when the PKCS15 emulator starts up, it /always/ fetches /all/ objects on the card, no matter what sort of operation I am performing.
>
> Oh... Hey, wait a sec... /AH-HAH/! I think I see where the confusion is coming form. The PIV app <https://developers.yubico.com/yubico-piv-tool/YubiKey_NEO_PIV_introduction.html> on the YubiKey NEO
> <https://www.yubico.com/products/yubikey-hardware/yubikey-neo/> (which I'm using for prototyping and happens to have the other certs and keys present) does not differentiate between NFC and USB,
> allowing /ALL/ operations to be performed via NFC (which is really unfortunate from a security standpoint). For a real, compliant PIV card (which is what I intend to use for this application), this
> wouldn't be a problem because the only objects that could be fetched would be the CHUID and the CAC. This would likely make the delay between credential presentation to the reader and the door
> unlocking much more acceptable in practice.
>
> In any case, I don't think that allowing the built-in PKCS15 caching mechanism to work properly would break anything, for two reasons:
>
>  1. You must opt in to use the file cache by setting |use_file_cache|. By default it is turned off.
>  2. |use_file_cache| does not automatically cache objects. If you want a card to be cached, you must explicitly do so.
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/377#issuecomment-75345458>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by darconeous** [comment]
Oh, nice! I'll give that a try.

**Comment by dengert** [comment]
There are two options, one says treat the card as if it is in a contactless reader.
The second is to  test the contactless code without a contactless reader.

(I purchased through Amazon a ACR122U-A9 this week, so now have a good contactless reader.)


On 2/20/2015 7:55 PM, Robert Quattlebaum wrote:
>     I am working on some mods for contactless access to PIV, that should speed up processing by only trying to load the objects that are allowed in contactless mode. I am also trying to accommodate
>     not standard PIV like devices that violate this standard. The opensc.conf would have to be configured to violate the contactless mode.
>
> This sounds great, especially if it could be configured to always treat PIV cards as "contactless" and only fetch the contactless subset. (In my application, this would be the case)
>
>     A missing piece at the moment is how to have pcscd to pass in the fact that a reader contactless.
>
>  From a PCSC perspective, you'd probably have to introspect into the description of the reader and differentiate between contact and contactless readers (In the case where it supports both I guess
> you'd assume contact-only). Kinda yucky, but doable.
>
>     The security is really in the card, but knowing of it is a contactless reader would improve performance.
>
> Certainly better than nothing!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/377#issuecomment-75349827>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by darconeous** [comment]
I was thinking about adding a new command to one of the `?-tool` commands: `--read-and-check-certificate`. The command would output the X.509 certificate for the indicated object, but only after challenging the token to verify that it has the private key associated with that certificate (asking for a PIN, if necessary).

Not exactly sure how the semantics would all work, which tool command it would be best suited for, or even if it is a reasonable name... It would be up to the caller to check the returned certificate is otherwise valid (chain-of-trust, revocation, etc).

**Comment by dengert** [comment]
#462 committed, that fixes the original double reads for  cards using T=1.
https://github.com/dengert/OpenSC/compare/piv-contactless
is not the way to go, and will be rewritten. The reader needs to be queried using part10 to findout if it is a contactless reader. (This may be across a network via rdesktop or a virtual machine.) Only then would the code not try and quere for objects that are not to be supported over contactless. But this interferres with the NEO that does not meet the NIST specs.

**Comment by frankmorgner** [comment]
@darconeous did you already submit a PR for https://github.com/OpenSC/OpenSC/commit/0ffad3f3bb9c893ab5a06998c7c7f0d571e7918d ? If so, we could close this issue (and accept the PR, as @dengert said).

**Comment by frankmorgner** [comment]
closing this as wontfix
